src/ast/nodes/ObjectExpression.ts

```diff
@@ -25,89 +24,151 @@ import Literal from './Literal';
 import * as NodeType from './NodeType';
 import Property from './Property';
 import { ExpressionEntity } from './shared/Expression';
-import { Node, NodeBase } from './shared/Node';
+import { NodeBase } from './shared/Node';
+import SpreadElement from './SpreadElement';
 
-const PROPERTY_KINDS_READ = ['init', 'get'];
-const PROPERTY_KINDS_WRITE = ['init', 'set'];
-
-export function isObjectExpression(node: Node): node is ObjectExpression {
-	return node.type === NodeType.ObjectExpression;
+interface PropertyMap {
+	[key: string]: {
+		exactMatchWrite: Property | null;
+		propertiesSet: Property[];
+		exactMatchRead: Property | null;
+		propertiesRead: (Property | SpreadElement)[];
+	};
 }
 
 export default class ObjectExpression extends NodeBase {
 	type: NodeType.tObjectExpression;
-	properties: Property[];
+	properties: (Property | SpreadElement)[];
 
+	// Caching and deoptimization:
+	// We collect deoptimization information if we can resolve a computed property access
+	private propertyMap: PropertyMap | null;
+	private unmatchablePropertiesRead: (Property | SpreadElement)[] | null;
+	private unmatchablePropertiesWrite: Property[] | null;
 	private reassignedPaths: { [key: string]: true };
 	private hasUnknownReassignedProperty: boolean;
+	private expressionsToBeDeoptimized: { [key: string]: DeoptimizableEntity[] };
+
+	bind() {
+		super.bind();
+		if (this.propertyMap === null) this.buildPropertyMap();
+	}
+
+	// We could also track this per-property but this would quickly become much more complex
+	deoptimize() {
+		if (!this.hasUnknownReassignedProperty) this.reassignAllProperties();
```

Comment:
It does seem a little unfortunate that the following:

```js
var obj = {
  x: 'x',
  y: 'y'
};
var test = false;

function asdf () {
  return obj;
}

if (asdf().x === 'x')
  console.log('yes');

if (obj.x === 'x')
  console.log('true');

function updateTest () {
  test = true;
}
```

loses all the analysis benfits if the return statement is just changed to `test ? obj : obj`.

Reply:
I might give it a try to improve a little here

Reply:
No, it seems with the return expression handling, fixing this would be a little more involved. This will probably need to wait for SSA.